commit 6f71139576710f2540b5971696788f73fa207775
Author: Patrick Mooney <pmooney@pfmooney.com>
Date:   2019-05-31T23:06:36+00:00 (4 months ago)
    
    OS-7812 viona bungles refcount during ipf hook
    Reviewed by: Ryan Zezeski <rpz@joyent.com>
    Reviewed by: Jason King <jbk@joyent.com>
    Approved by: Jason King <jbk@joyent.com>

diff --git a/usr/src/uts/i86pc/io/viona/viona.c b/usr/src/uts/i86pc/io/viona/viona.c
index 94b1b9ce2d..bbcb970b22 100644
--- a/usr/src/uts/i86pc/io/viona/viona.c
+++ b/usr/src/uts/i86pc/io/viona/viona.c
@@ -2990,8 +2990,22 @@ viona_tx(viona_link_t *link, viona_vring_t *ring)
 			goto drop_hook;
 		}
 
-		if (dp != NULL)
+		if (dp != NULL) {
 			dp->d_ref--;
+
+			/*
+			 * It is possible that the hook(s) accepted the packet,
+			 * but as part of its processing, it issued a pull-up
+			 * which released all references to the desb.  In that
+			 * case, go back to acting like the packet is entirely
+			 * copied (which it is).
+			 */
+			if (dp->d_ref == 1) {
+				dp->d_cookie = 0;
+				dp->d_ref = 0;
+				dp = NULL;
+			}
+		}
 	}
 
 	/*
